File: app/controllers/exif_data_controller.rb
class ExifDataController < ApplicationController
  # GET /exif_data
  # GET /exif_data.json
  def index
    @exif_data = ExifDatum.all

    respond_to do |format|
      format.html # index.html.erb
      format.json { render json: @exif_data }
    end
  end

  # GET /exif_data/1
  # GET /exif_data/1.json
  def show
    @exif_datum = ExifDatum.find(params[:id])

    respond_to do |format|
      format.html # show.html.erb
      format.json { render json: @exif_datum }
    end
  end

  # GET /exif_data/new
  # GET /exif_data/new.json
  def new
    @exif_datum = ExifDatum.new

    respond_to do |format|
      format.html # new.html.erb
      format.json { render json: @exif_datum }
    end
  end

  # GET /exif_data/1/edit
  def edit
    @exif_datum = ExifDatum.find(params[:id])
  end

  # POST /exif_data
  # POST /exif_data.json
  def create
    @exif_datum = ExifDatum.new(params[:exif_datum])

    respond_to do |format|
      if @exif_datum.save
        format.html { redirect_to @exif_datum, notice: 'Exif datum was successfully created.' }
        format.json { render json: @exif_datum, status: :created, location: @exif_datum }
      else
        format.html { render action: "new" }
        format.json { render json: @exif_datum.errors, status: :unprocessable_entity }
      end
    end
  end

  # PUT /exif_data/1
  # PUT /exif_data/1.json
  def update
    @exif_datum = ExifDatum.find(params[:id])

    respond_to do |format|
      if @exif_datum.update_attributes(params[:exif_datum])
        format.html { redirect_to @exif_datum, notice: 'Exif datum was successfully updated.' }
        format.json { head :no_content }
      else
        format.html { render action: "edit" }
        format.json { render json: @exif_datum.errors, status: :unprocessable_entity }
      end
    end
  end

  # DELETE /exif_data/1
  # DELETE /exif_data/1.json
  def destroy
    @exif_datum = ExifDatum.find(params[:id])
    @exif_datum.destroy

    respond_to do |format|
      format.html { redirect_to exif_data_url }
      format.json { head :no_content }
    end
  end
end

File: Sources/App/Controllers/ExifDataController.swift
import Vapor
import Fluent

struct ExifDataController {
    /// GET /exif_data — tag/value counts with filter hashes (same as browse list).
    func index(req: Request) async throws -> View {
        let exif = try await BrowseHelpers.tagCounts(on: req.db)
        struct Context: Content { let exif: [ExifTagCount] }
        return try await req.view.render("exif_data/index", Context(exif: exif))
    }

    /// GET /exif_data/new
    func new(req: Request) async throws -> View {
        struct Context: Content { let datum: ExifDatum? }
        return try await req.view.render("exif_data/new", Context(datum: nil))
    }

    struct ExifForm: Content {
        var image_id: Int?
        var tag: String?
        var value: String?
    }

    /// POST /exif_data — create.
    func create(req: Request) async throws -> Response {
        let form = try req.content.decode(ExifForm.self)
        guard let imageID = form.image_id else {
            throw Abort(.badRequest, reason: "image_id is required")
        }
        let datum = ExifDatum(imageID: imageID, tag: form.tag, value: form.value)
        try await datum.save(on: req.db)
        return req.redirect(to: "/exif_data/\(datum.id ?? 0)")
    }

    /// GET /exif_data/:id — show.
    func show(req: Request) async throws -> View {
        guard let id = req.parameters.get("id", as: Int.self),
              let datum = try await ExifDatum.query(on: req.db)
                .filter(\.$id == id)
                .with(\.$image)
                .first() else {
            throw Abort(.notFound)
        }
        struct Context: Content { let datum: ExifDatum }
        return try await req.view.render("exif_data/show", Context(datum: datum))
    }

    /// GET /exif_data/:id/edit
    func edit(req: Request) async throws -> View {
        guard let id = req.parameters.get("id", as: Int.self),
              let datum = try await ExifDatum.query(on: req.db)
                .filter(\.$id == id)
                .with(\.$image)
                .first() else {
            throw Abort(.notFound)
        }
        struct Context: Content { let datum: ExifDatum }
        return try await req.view.render("exif_data/edit", Context(datum: datum))
    }

    /// POST /exif_data/:id — update.
    func update(req: Request) async throws -> Response {
        guard let id = req.parameters.get("id", as: Int.self),
              let datum = try await ExifDatum.find(id, on: req.db) else {
            throw Abort(.notFound)
        }
        let form = try req.content.decode(ExifForm.self)
        if let imageID = form.image_id { datum.$image.id = imageID }
        datum.tag = form.tag
        datum.value = form.value
        try await datum.save(on: req.db)
        return req.redirect(to: "/exif_data/\(id)")
    }

    /// POST /exif_data/:id/delete — destroy.
    func delete(req: Request) async throws -> Response {
        guard let id = req.parameters.get("id", as: Int.self),
              let datum = try await ExifDatum.find(id, on: req.db) else {
            throw Abort(.notFound)
        }
        try await datum.delete(on: req.db)
        return req.redirect(to: "/exif_data")
    }
}
